tools/nodeset_compiler/backend_open62541_nodes.py

```diff
@@ -96,65 +169,81 @@ def generateVariableNodeCode(node, nodeset, encode_binary_size):
         code.append("attr.arrayDimensions = &arrayDimensions[0];")
 
-    if node.dataType is not None:
-        if isinstance(node.dataType, NodeId) and node.dataType.ns == 0 and node.dataType.i == 0:
-            #BaseDataType
-            dataTypeNode = nodeset.nodes[NodeId("i=24")]
-            dataTypeNodeOpaque = nodeset.nodes[NodeId("i=24")]
-        else:
-            dataTypeNodeOpaque = nodeset.getDataTypeNode(node.dataType)
-            dataTypeNode = nodeset.getBaseDataType(nodeset.getDataTypeNode(node.dataType))
-
-        if dataTypeNode is not None:
-            code.append("attr.dataType = %s;" % generateNodeIdCode(dataTypeNodeOpaque.id))
-
-            if dataTypeNode.isEncodable():
-                if node.value is not None:
-                    [code1, codeCleanup1, codeGlobal1] = generateValueCode(node.value, nodeset.nodes[node.id], nodeset)
-                    code += code1
-                    codeCleanup += codeCleanup1
-                    codeGlobal += codeGlobal1
-                    # #1978 Variant arrayDimensions are only required to properly decode multidimensional arrays
-                    # (valueRank >= 2) from data stored as one-dimensional array of arrayLength elements.
-                    # One-dimensional arrays are already completely defined by arraylength attribute so setting
-                    # also arrayDimensions, even if not explicitly forbidden, can confuse clients
-                    if node.valueRank > 1 and len(node.arrayDimensions) == node.valueRank:
-                        code.append("attr.value.arrayDimensionsSize = attr.arrayDimensionsSize;")
-                        code.append("attr.value.arrayDimensions = attr.arrayDimensions;")
+    if node.dataType is None:
+        # Inherit the datatype from the HasTypeDefinition reference, as stated in the OPC UA Spec:
+        # 6.4.2
+        # "Instances inherit the initial values for the Attributes that they have in common with the
+        # TypeDefinitionNode from which they are instantiated, with the exceptions of the NodeClass and
+        # NodeId."
+        setNodeDatatypeRecursive(node, nodeset)
+        code.append("/* DataType inherited */")
+
+    dataTypeNode = nodeset.getBaseDataType(nodeset.getDataTypeNode(node.dataType))
+
+    if dataTypeNode is None:
+        raise RuntimeError("Cannot get BaseDataType for dataType : " + str(node.dataType) + " of node " + node.browseName.name + " " + str(node.id))
+
+    code.append("attr.dataType = %s;" % generateNodeIdCode(node.dataType))
+
+    if dataTypeNode.isEncodable():
+        if node.value is not None:
+            [code1, codeCleanup1, codeGlobal1] = generateValueCode(node.value, nodeset.nodes[node.id], nodeset)
+            code += code1
+            codeCleanup += codeCleanup1
+            codeGlobal += codeGlobal1
+            if node.valueRank is not None and node.valueRank > 0 and len(node.arrayDimensions) == node.valueRank and len(node.value.value) > 0:
```

Comment:
I think there is a problem, to my understanding, **the fix done in #1978 was lost when you force-pushed this pull request in master**.
To re-introduce the lost fix, if I’m not wrong, in the file “backend_open62541_nodes.py” (about in line 194):
if node.valueRank is not None and node.valueRank > 0 and len(node.arrayDimensions) == node.valueRank and len(node.value.value) > 0:
it should be changed to “…node.valueRank > 1...”, i.e. > 1 instead of > 0.
In our case the problem shows in two OPC-UA Client (“Unified Automation UaExpert” and “Integration Objects OPC UA Explorer”):
**when trying to do a Method Call, no Input / Output Arguments are shown (UA Explorer tells they’re Opc.Ua.Matrix instead of array**)

Reply:
Thanks for the hint! 

Can you please create a PR (or at least a new Issue) and also add the comment form #1978 to your PR.

Is there a way to add a unit test to avoid this in the future?

(Discussions in closed PRs and issues may get lost quite easily)

Reply:
OK, I forked, did the fix, added the comment and unit test, and created PR #2853 .